File: bestshuffle.rb
# http://rosettacode.org/wiki/Best_shuffle

def best_shuffle(word,shuffled_word=[])
  chars = word.split(//)
  counter = 0; compare = false
  until compare == true
    counter = 0; shuffled_word = word.split(//).shuffle
    chars.each_index {|i| counter += 1 if shuffled_word[i] == chars[i]}
    compare = true if counter == 0 || chars.count < 2 || (chars.count > chars.uniq.count * 3 && counter < chars.count)
  end
  puts "#{word}, #{shuffled_word.join}, #{counter}"
end

["abracadabra", "seesaw", "elk", "grrrrrr", "up", "a"].each {|word| best_shuffle(word)}

File: best_shuffle.pl
#!usr/bin/perl
use strict;
use warnings;

sub shuffle {
  my @array = @_;
  my $i = scalar @array;
  while (--$i) {
  my $j = int rand( $i + 1 );
  @array[$i,$j] = @array[$j,$i];
  }
  return @array;
}

sub uniq {
  my @list = @_;
  my @uniq_list = ();
  for (my $i = 0; $i < scalar @list; $i++) {
    push (@uniq_list,$list[$i]) unless grep {$_ eq $list[$i]} @uniq_list
  }
  return scalar @uniq_list;
}

sub compare {
  my $word = shift;
  my @chars = split(//,$word);
  my @shuffled = @chars;
  my $best_shuffle = 0;
  my $counter = 0;
  until ($best_shuffle) {
    @shuffled = shuffle(@chars);
    $counter = 0;
    for (my $i = 0; $i < scalar @chars; $i++) {
      $counter++ if $chars[$i] eq $shuffled[$i];
    }
    $best_shuffle = 1 if $counter == 0;
    $best_shuffle = 1 if scalar @chars < 2;
    $best_shuffle = 1 if scalar @chars > uniq(@chars) * 3 && $counter < scalar @chars;
  }
  my $shuffled_word = join("",@shuffled);
  print "$word, $shuffled_word, ($counter)\n";
}

my @words = ("abracadabra", "seesaw", "elk", "grrrrrr", "up", "a");
foreach (@words) {
  compare($_);
}
